### PR TITLE
Store Refactor

### PR DESCRIPTION
**Pull request by InTheCloudDan**
This PR refactors the Store to expose a public `flagMetadata` that other classes can consume. It also creates a new event emitter that those classes can listen to when there is a flag change. This is different that current implementation which exposes the ldclient `On` listener.

It also adds error handling into API class.

### CODE REVIEW

File: src/providers/flagsView.ts [hunk 1]
```diff
@@ -56,6 +58,20 @@ export class LaunchDarklyTreeViewProvider implements vscode.TreeDataProvider<Fla
...
+	private load(flags: Object): Array<FlagNode> {
+		const parsedFlags: Array<FlagNode> = [];
+		// clear all existing nodes
+		this.flagNodes = [];
+		Object.keys(flags).map((key, idx, arr) => {
+			this.flagNodes.push(this.flagToValues(flags[key]));
+			if (idx == arr.length - 1) {
+				console.log('load refresh');
+				this.refresh();
+			}
```
Comment (atrakh) [inline src/providers/flagsView.ts:70]: why not do this outside the map to avoid a conditional block?

File: src/flagStore.ts [hunk 2]
```diff
@@ -64,13 +66,43 @@ export class FlagStore {
...
+		var CronJob = cron.CronJob;
+		var job = new CronJob(
+			exp,
+			async () => {
+				this.updateFlags();
+			},
+			null,
+			true,
+		);
+		job.start();
```
[the remark this answers is not in the document]
Comment (InTheCloudDan) [reply src/flagStore.ts:102]: cron package seemed much more reliable and flexible for users. Though with just minutes it might be possible. I'll first look into just converting minutes to a cron expression and get everything working there. As a future much smaller refactoring I can look to remove this package.

File: src/providers.ts [hunk 3]
```diff
@@ -164,11 +190,11 @@ class LaunchDarklyCompletionItemProvider implements CompletionItemProvider {
 }
 
 const openFlagInBrowser = async (config: Configuration, flagKey: string, flagStore: FlagStore) => {
-	const { flag } = await flagStore.getFeatureFlag(flagKey);
+	const flag = await flagStore.getFeatureFlag(flagKey);
 
 	// Default to first environment
-	let env: FeatureFlagConfig = Object.values(flag.environments)[0];
-	let sitePath = env._site.href;
+	let env = <Array<FeatureFlagConfig>>Object.values(flag.environments);
```
[the remark this answers is not in the document]
Comment (InTheCloudDan) [reply src/providers.ts]: I had to add a `<unknown>` here if you want to double check the updates.

File: README.md [hunk 4]
```diff
@@ -23,9 +23,9 @@ This extension contributes the following additional settings:
...
 | `launchdarkly.streamUri`          |                         The LaunchDarkly stream uri to be used. Optional.                          | `https://stream.launchdarkly.com` |
 | `launchdarkly.enableHover`        |                  Enables flag info to be displayed on hover of a valid flag key.                   |                            `true` |
-| `launchdarkly.enableAutocomplete` |                                  Enable flag key |autocompletion.                                   |                            `true` |
+| `launchdarkly.enableAutocomplete` |                                  Enable flag key autocompletion.                                   |                            `true` |
 | `launchdarkly.enableFlagExplorer`	|           Show all of the feature flags for the configured environment within the project.         |                            `true` |
-| `launchdarkly.refreshCron`		|		Cron Expression for how often to refresh feature flag metadata via API. | `* */2 * * *` |
+| `launchdarkly.refreshRate`		|		How often in minutes to refresh feature flag metadata via API. If `0` flags do not update. | `120` |
```
Comment (atrakh) [inline README.md:28]: > If 0 flags do not update

Is this true only for global flag data, or configs too?

File: package.json [hunk 5]
```diff
@@ -67,10 +67,10 @@
...
 					"description": "LaunchDarkly SDK key. OBSOLETE: Run the 'LaunchDarkly: Configure' command instead."
 				},
-				"launchdarkly.refreshCron": {
-					"type": "string",
-					"default": "* */2 * * *",
-					"description": "Cron Expression for how often to query LaunchDarkly API for metadata."
+				"launchdarkly.refreshRate": {
+					"type": "number",
+					"default": 120,
+					"description": "How often in minutes to query LaunchDarkly API for updated metadata."
```
Comment (atrakh) [inline package.json:73]: nit: README documentation and description in code should be the same

File: src/configuration.ts [hunk 6]
```diff
@@ -11,6 +11,7 @@ export class Configuration {
 	sdkKey = '';
 	project = '';
 	env = '';
+	refreshRate = 0;
```
Comment (atrakh) [inline src/configuration.ts:14]: The default is 120 right?

File: src/flagStore.ts [hunk 7]
```diff
@@ -64,13 +65,37 @@ export class FlagStore {
...
+				}
+			}
 		} catch (err) {
 			this.rejectLDClient();
 			console.error(err);
 		}
 	}
 
-	async on(event: string, cb: FlagUpdateCallback) {
+	private async cron(exp: number) {
```
Comment (atrakh) [inline src/flagStore.ts:90]: Let's give this a name that might give the reader a better idea of what it does. Maybe something like `startGlobalFlagUpdateTask(interval: number)`

File: src/models.ts [hunk 8]
```diff
@@ -320,12 +319,16 @@ export class Defaults {
 	}
 }
 
+class EnvironmentMap {
+	[key: string]: FeatureFlagConfig | FlagConfiguration;
```
Comment (atrakh) [inline src/models.ts:323]: What's the difference between `FeatureFlagConfig` and `FlagConfiguration`? Is one of these the API result and the other the stream result? Perhaps we could discard the API results since we're always processing updates from the stream? Ideally we would minimize the amount of redundant types and data

File: src/providers.ts [hunk 9]
```diff
@@ -31,7 +33,31 @@ const LD_MODE: DocumentFilter = {
...
+	var flagStore;
+
+	try {
+		const flags = await api.getFeatureFlags(config.project, config.env);
+		const arrayToObject = (array: Array<FeatureFlag>) =>
+			array.reduce((obj: { [key: string]: FeatureFlag }, item): { [key: string]: FeatureFlag } => {
+				obj[item.key] = item;
+				return obj;
+			}, {});
+		let intFlags = arrayToObject(flags);
```
Comment (atrakh) [inline src/providers.ts:46]: what are intFlags?

File: src/providers.ts [hunk 10]
```diff
@@ -31,7 +33,31 @@ const LD_MODE: DocumentFilter = {
 	scheme: 'file',
 };
 
-export function register(ctx: ExtensionContext, config: Configuration, flagStore: FlagStore, api: LaunchDarklyAPI) {
+export async function register(ctx: ExtensionContext, config: Configuration, api: LaunchDarklyAPI) {
+	var flagStore;
+
+	try {
+		const flags = await api.getFeatureFlags(config.project, config.env);
+		const arrayToObject = (array: Array<FeatureFlag>) =>
```
Comment (atrakh) [inline src/providers.ts:41]: Looks like we've declared this method in multiple places. Since it gets a little funky with typescript, maybe we could use lodash's [`keyBy`](https://lodash.com/docs/4.17.15#keyBy): 

```
keyBy(flags, 'key')
```

File: src/models.ts (hunk 8 above)
Comment (InTheCloudDan) [reply src/models.ts:323]: Yes that's it exactly. It might be possible to discard it but I'm not 100% certain at this time. With this refactor overall we are drastically reducing redundant data. We are merging the targeting with the API Flag Summary. So there is no duplication of data at that point.

File: src/providers.ts (hunk 10 above)
Comment (InTheCloudDan) [reply src/providers.ts:41]: that's awesome! Worked like a charm.

File: src/providers.ts [hunk 11]
```diff
@@ -193,8 +193,8 @@ const openFlagInBrowser = async (config: Configuration, flagKey: string, flagSto
 	const flag = await flagStore.getFeatureFlag(flagKey);
 
 	// Default to first environment
-	let env = <Array<FeatureFlagConfig>>Object.values(flag.environments);
-	let sitePath = env[0]._site.href;
+	let env = <FeatureFlagConfig>(<unknown>Object.values(flag.environments));
```
[the remark this answers is not in the document]
Comment (InTheCloudDan) [reply src/providers.ts:196]: Based on the previous conversation I thought you wanted that cleaned up? We can chat about this on Monday.

File: README.md [hunk 12]
```diff
@@ -25,6 +25,7 @@ This extension contributes the following additional settings:
 | `launchdarkly.enableHover`        |                  Enables flag info to be displayed on hover of a valid flag key.                   |                            `true` |
 | `launchdarkly.enableAutocomplete` |                                  Enable flag key autocompletion.                                   |                            `true` |
 | `launchdarkly.enableFlagExplorer`	|           Show all of the feature flags for the configured environment within the project.         |                            `true` |
+| `launchdarkly.refreshRate`		|		How often in minutes to refresh feature flag metadata via API. If `0` global flag metadata does not update. Max is 1440(1 Day). | `120` |
```
Comment (atrakh) [inline README.md:28]:
```suggestion
| `launchdarkly.refreshRate`		|		How often in minutes to refresh feature flag metadata via API. If `0` global flag metadata does not update. Max is 1440 (1 Day). | `120` |
```

File: package.json [hunk 13]
```diff
@@ -66,6 +66,11 @@
 					"type": "string",
 					"default": "",
 					"description": "LaunchDarkly SDK key. OBSOLETE: Run the 'LaunchDarkly: Configure' command instead."
+				},
+				"launchdarkly.refreshRate": {
+					"type": "number",
+					"default": 120,
+					"description": "How often in minutes to refresh feature flag metadata via API. If `0` global flag metadata does not update. Max is 1440(1 Day)."
```
Comment (atrakh) [inline package.json:73]:
```suggestion
					"description": "How often in minutes to refresh feature flag metadata via API. If `0` global flag metadata does not update. Max is 1440 (1 Day)."
```

File: src/flagStore.ts [hunk 14]
```diff
@@ -64,13 +65,36 @@ export class FlagStore {
...
+				} catch (err) {
+					console.error('Failed to update LaunchDarkly flag store.', err);
+				}
+			});
+			if (this.config.refreshRate) {
+				if (this.config.validateRefreshInterval(this.config.refreshRate)) {
+					this.startGlobalFlagUpdateTask(this.config.refreshRate);
+				} else {
+					window.showErrorMessage(
+						`Invalid Refresh time(in Minutes): '${this.config.refreshRate}'. 0 is off, up to 1440 for one day.`,
```
Comment (atrakh) [inline src/flagStore.ts:80]:
```suggestion
						`Invalid Refresh time (in Minutes): '${this.config.refreshRate}'. 0 is off, up to 1440 for one day.`,
```

File: src/flagStore.ts [hunk 15]
```diff
@@ -129,8 +161,8 @@ export class FlagStore {
 		};
 	}
 
-	getFeatureFlag(key: string): Promise<FlagWithConfiguration | null> {
-		let flag = this.flagMetadata[key];
+	getFeatureFlag(key: string): Promise<FeatureFlag> {
+		let flag = JSON.parse(JSON.stringify(this.flagMetadata[key]));
```
[the remark this answers is not in the document]
Comment (atrakh) [reply src/flagStore.ts:158]: repinging this comment because I'm not sure if it was resolved

File: src/flagStore.ts [hunk 16]
```diff
@@ -142,21 +167,82 @@ export class FlagStore {
...
+				const retFlag = await this.mergeFlag(flag, res);
+				resolve(retFlag);
 			});
 		});
 	}
 
-	allFlags(): Promise<FlagConfiguration[]> {
+	async allFlags(): Promise<Record<string, FeatureFlag>> {
+		await this.ldClient;
+		const flagDeepClone = JSON.parse(JSON.stringify(this.flagMetadata));
```
Comment (atrakh) [inline src/flagStore.ts:191]: Same as the other comment - we shouldn't need to `JSON.parse(JSON.stringify())` anything

File: src/providers.ts [hunk 17]
```diff
@@ -122,16 +137,16 @@ class LaunchDarklyHoverProvider implements HoverProvider {
...
-						const env = data.flag.environments[this.config.env];
+						const env = <FeatureFlagConfig>data.environments[this.config.env];
 						const sitePath = env._site.href;
 						const browserUrl = url.resolve(this.config.baseUri, sitePath);
-						const hover = generateHoverString(data.flag, data.config, browserUrl);
+						const hover = generateHoverString(data, this.config.env, browserUrl);
 						resolve(new Hover(hover));
 						return;
+					} else {
+						console.log('no data');
```
Comment (atrakh) [inline src/providers.ts:149]: Is this log relevant to the client (user)?

File: src/providers.ts [hunk 18]
```diff
@@ -171,10 +186,10 @@ class LaunchDarklyCompletionItemProvider implements CompletionItemProvider {
 }
 
 const openFlagInBrowser = async (config: Configuration, flagKey: string, flagStore: FlagStore) => {
-	const { flag } = await flagStore.getFeatureFlag(flagKey);
+	const flag = await flagStore.getFeatureFlag(flagKey);
 
 	// Default to first environment
-	let env: FeatureFlagConfig = Object.values(flag.environments)[0];
+	let env = <FeatureFlagConfig>flag.environments[Object.keys(flag.environments)[0]];
```
Comment (atrakh) [inline src/providers.ts:192]: something like this might work
```suggestion
	let env = <FeatureFlagConfig>flag.environments.values().next().value;
```

Comment (InTheCloudDan) [reply src/providers.ts:192]: That one gives me an error:
```This expression is not callable.
  No constituent of type 'FeatureFlagConfig | FlagConfiguration' is callable.```

Comment (atrakh) [reply src/providers.ts:192]: Ah ok, we can do without this change then

File: src/flagStore.ts (hunk 16 above)
Comment (InTheCloudDan) [reply src/flagStore.ts:191]: Without the JSON stringify I'm getting an error like this:

```rejected promise not handled within 1 second: TypeError: Cannot read property 'dano' of undefined
extensionHostProcess.js:976
stack trace: TypeError: Cannot read property 'dano' of undefined
	at FlagStore.mergeFlag (/Users/danielobrien/Projects/ld-vscode/dist/extension.js:52184:62)
	at FlagStore.<anonymous> (/Users/danielobrien/Projects/ld-vscode/dist/extension.js:52154:44)
	at Generator.next (<anonymous>)
	at /Users/danielobrien/Projects/ld-vscode/dist/extension.js:51967:71```

line 52184 in extension.js is:
```                const retFlag = yield this.mergeFlag(flag, res);``` which is around line 183 https://github.com/launchdarkly/ld-vscode/blob/storeRefactor/src/flagStore.ts#L183

Comment (atrakh) [reply src/flagStore.ts:191]: This appears to be due to some incorrect downstream assumptions of what `flagMetadata` (the FlagMap) type should look like. The roundabout stringify -> parse forces loss of type data - returning a plain js object.

The `mergeFlag` and `mergeAll` methods look pretty suspect in how they reassign values to an object. Would recommend doing some debugging in that part to make sure the flag data is being merged as expected

Comment (atrakh) [reply src/flagStore.ts:191]: We might have briefly discussed this idea in a call, but I never wrote it down -- one way to avoid the complexity of merging flag configs would be to only use the API results for storing the global flag metadata, without attempting to merge environment configs. This approach should be ok because the stream guarantees up-to-date configuration data.

Comment (InTheCloudDan) [reply src/flagStore.ts]: the reason I had to merge it was for things like the Treeview, to give them a clean API to work with. I was able to remove the Stringify! Let me know what you think.

**Comment by InTheCloudDan** [comment]
This was overcomplicating the issue and trying to hide where the data is coming from. Going to close out and just use the necessary parts for a cleaner PR.
